examples/Custom_Hooks/probability_of_improvement.py

```diff
@@ -0,0 +1,219 @@
+## Example for calculating and plotting the Probability of Improvement
+
+# This example demonstrates how to use the
+# {func}`register_hooks <baybe.utils.basic.register_hooks>` utility
+# to calculate and analyse the probability of improvement (PI) evolving during iterations:
+# * We define a hook that is compatible with the general
+#   {meth}`RecommenderProtocol.recommend <baybe.recommenders.base.RecommenderProtocol.recommend>`
+#   interface,
+# * attach it to a recommender,
+# * and watch it take action.
+
+
+### Imports
+
+import os
+import warnings
+from pathlib import Path
+from types import MethodType
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import torch
+from botorch.test_functions.synthetic import Hartmann
+from matplotlib.axes import Axes
+from matplotlib.collections import PolyCollection
+from matplotlib.figure import Figure
+from scipy.stats import gaussian_kde
+
+from baybe.acquisition import ProbabilityOfImprovement
+from baybe.campaign import Campaign
+from baybe.objectives.base import Objective
+from baybe.objectives.single import SingleTargetObjective
+from baybe.parameters import NumericalDiscreteParameter
+from baybe.recommenders import (
+    BotorchRecommender,
+    RandomRecommender,
+    TwoPhaseMetaRecommender,
+)
+from baybe.searchspace import SearchSpace, SearchSpaceType
+from baybe.surrogates import GaussianProcessSurrogate
+from baybe.targets import NumericalTarget
+from baybe.utils.basic import register_hooks
+from baybe.utils.botorch_wrapper import botorch_function_wrapper
+from baybe.utils.dataframe import to_tensor
+
+### Parameters for a full simulation loop
+
+# For the full simulation, we need to define some additional parameters. These are the number of experiments to be conducted per run, the batch size, the dimension and the points per dimension.
+
+SMOKE_TEST = "SMOKE_TEST" in os.environ
+N_DOE_ITERATIONS = 3 if SMOKE_TEST else 7
+BATCH_SIZE = 2
+DIMENSION = 3
+POINTS_PER_DIM = 4
+
+### Setup
+
+# We start by initializing the global `poi_list` where we will store the PI from each recommend iteration:
+
+pi_list: list[torch.Tensor] = []
```

Comment:
Not a big fan of including the type of a variable in its name (again, for refactorability but also because the specific type of collection – in this case a list – is completely irrelevant to make the point of the example). Can you use a more meaningful name? Like `pi_per_iteration` or similar?

Reply:
Sure! I refactored the name based on your suggestion 👍